### PR TITLE
Migrate Jmockit VerificationsInOrder to Mockito

### PR DESCRIPTION
**Pull request by shivanisky**
## What's changed?
Add feature to migrate Jmockit VerificationsInOrder to mockito https://javadoc.io/doc/org.jmockit/jmockit/latest/mockit/VerificationsInOrder.html
https://www.javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/InOrder.html

## What's your motivation?
- https://github.com/jmockit/jmockit1/issues/729#issuecomment-2406766342

## Anyone you would like to review specifically?
@sambsnyd  @timtebeek @tinder-dthomson as FYI

## Any additional context
Only JMockit Expectations Delegate left to migrate.

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java [hunk 1]
```diff
@@ -29,21 +29,22 @@
...
-import static org.openrewrite.java.testing.jmockit.JMockitBlockType.NonStrictExpectations;
+import static org.openrewrite.java.testing.jmockit.JMockitBlockType.*;
 
 class JMockitBlockRewriter {
 
     private static final String WHEN_TEMPLATE_PREFIX = "when(#{any()}).";
     private static final String VERIFY_TEMPLATE_PREFIX = "verify(#{any()}";
     private static final String VERIFY_NO_INTERACTIONS_TEMPLATE_PREFIX = "verifyNoMoreInteractions(";
+    private static final String VERIFY_IN_ORDER_TEMPLATE_PREFIX = "inOrder(";
+    //private static final String VERIFY_IN_ORDER_TEMPLATE_PREFIX = "InOrder inOrder = inOrder(";
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java:40]:
```suggestion
```

File: src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java [hunk 2]
```diff
@@ -128,11 +129,16 @@ J.Block rewriteMethodBody() {
             removeBlock();
         }
 
+        ArrayList<Object> mocks = new ArrayList<>(uniqueMocks);
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java:132]:
```suggestion
        List<Object> mocks = new ArrayList<>(uniqueMocks);
```

File: src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java [hunk 3]
```diff
@@ -243,7 +253,7 @@ private void rewriteVerify(J.MethodInvocation invocation, @Nullable Expression t
         }
     }
 
-    private void rewriteFullVerify(List<Object> mocks) {
+    private void rewriteFullVerify(ArrayList<Object> mocks) {
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java:256]:
```suggestion
    private void rewriteFullVerify(List<Object> mocks) {
```

File: src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java [hunk 4]
```diff
@@ -257,6 +267,21 @@ private void rewriteFullVerify(List<Object> mocks) {
         }
     }
 
+    private void rewriteInOrderVerify(ArrayList<Object> mocks) {
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java:270]:
```suggestion
    private void rewriteInOrderVerify(List<Object> mocks) {
```

**Review by timtebeek — COMMENTED** [review]
Great to see you follow through with the penultimate feature needed to complete the JMockit migrations. I've done a very quick round of comments from GitHub; will need to look in more detail later.

File: src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java (hunk 1 above)
Comment (shivanisky) [reply src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java:40]: Ah yes - this is one issue I'd like to get input on
inOrder(mock) works, but
InOrder inOrder = inOrder(mock) causes error even though I have the right compilation options and the right import. Hence I am not able to get all the tests passing (have disabled for now).

Caused by: java.lang.IllegalArgumentException: Could not parse as Java
	at org.openrewrite.java.internal.template.JavaTemplateParser.lambda$compileTemplate$13(JavaTemplateParser.java:264)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at org.openrewrite.java.internal.template.JavaTemplateParser.compileTemplate(JavaTemplateParser.java:264)
	at org.openrewrite.java.internal.template.JavaTemplateParser.lambda$parseBlockStatements$9(JavaTemplateParser.java:176)
	at org.openrewrite.java.internal.template.JavaTemplateParser.lambda$cacheIfContextFree$14(JavaTemplateParser.java:287)
	at org.openrewrite.java.internal.template.JavaTemplateParser.cache(JavaTemplateParser.java:308)
	at org.openrewrite.java.internal.template.JavaTemplateParser.cacheIfContextFree(JavaTemplateParser.java:287)
	at org.openrewrite.java.internal.template.JavaTemplateParser.parseBlockStatements(JavaTemplateParser.java:171)
	at org.openrewrite.java.internal.template.JavaTemplateJavaExtension$1.lambda$visitBlock$1(JavaTemplateJavaExtension.java:112)
	at org.openrewrite.internal.ListUtils.lambda$flatMap$0(ListUtils.java:261)
	at org.openrewrite.internal.ListUtils.flatMap(ListUtils.java:205)
	at org.openrewrite.internal.ListUtils.flatMap(ListUtils.java:261)
	at org.openrewrite.java.internal.template.JavaTemplateJavaExtension$1.visitBlock(JavaTemplateJavaExtension.java:110)
	at org.openrewrite.java.internal.template.JavaTemplateJavaExtension$1.visitBlock(JavaTemplateJavaExtension.java:55)
	at org.openrewrite.java.tree.J$Block.acceptJava(J.java:838)
	at org.openrewrite.java.tree.J.accept(J.java:59)
	at org.openrewrite.TreeVisitor.visit(TreeVisitor.java:250)
	... 58 more

Comment (shivanisky) [reply src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java]: I updated branch with main and still seeing the same issue. Strangely it's all working for verifyNoMoreInteractions(mock1, mock2, ...) but doesn't work for InOrder inOrder = inOrder(mock1, mock 2...). If I do just inOrder(mock1, mock2...) it seems to work. This looks like a framework issue, unless I'm missing something here.

Also seeing many of these warnings:
23:18:03.671 [main] WARN org.openrewrite.java.isolated.ReloadableJava17Parser - warning: unknown enum constant Status.STABLE
23:18:03.673 [main] WARN org.openrewrite.java.isolated.ReloadableJava17Parser -   reason: class file for org.apiguardian.api.API$Status not found

Comment (shivanisky) [reply src/main/java/org/openrewrite/java/testing/jmockit/JMockitBlockRewriter.java]: Ah all good, that parse error went away - thanks, maybe all that was needed was to sync with main! Now just getting some strange formatting issue, just an empy new line between statements in test whenThreeBlocks, hence one test is failing. Not due to having 3 blocks, but due to having a very short InOrder block of 2 lines.

**Comment by timtebeek** [comment]
Looks like we're very close, with only one failure left on whitespace
```
JMockitVerificationsInOrderToMockitoTest > whenThreeBlocks() FAILED
    org.opentest4j.AssertionFailedError: [Unexpected result in "MyTest.java":
```
```diff
diff --git a/MyTest.java b/MyTest.java
index 99fe3e8..bd0a50f 100644
--- a/MyTest.java
+++ b/MyTest.java
@@ -18,11 +18,15 @@ 
         inOrder.verify(obj).wait();

         obj.wait();
+
         InOrder inOrder1 = inOrder(obj);
+
         inOrder1.verify(obj).wait();

         obj.wait(10L, 10);
+
         InOrder inOrder2 = inOrder(obj);
+
         inOrder2.verify(obj).wait(anyLong(), anyInt());
     }
 }
\ No newline at end of file
``` 
## expected: 
```java
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InOrder;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class MyTest {
    @Mock
    Object obj;

    void test() {
        obj.wait(10L, 10);
        obj.wait();
        InOrder inOrder = inOrder(obj);
        inOrder.verify(obj).wait(anyLong(), anyInt());
        inOrder.verify(obj).wait();

        obj.wait();
        InOrder inOrder1 = inOrder(obj);
        inOrder1.verify(obj).wait();

        obj.wait(10L, 10);
        InOrder inOrder2 = inOrder(obj);
        inOrder2.verify(obj).wait(anyLong(), anyInt());
    }
}
```
## but was:
```java
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InOrder;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class MyTest {
    @Mock
    Object obj;

    void test() {
        obj.wait(10L, 10);
        obj.wait();
        InOrder inOrder = inOrder(obj);
        inOrder.verify(obj).wait(anyLong(), anyInt());
        inOrder.verify(obj).wait();

        obj.wait();

        InOrder inOrder1 = inOrder(obj);

        inOrder1.verify(obj).wait();

        obj.wait(10L, 10);

        InOrder inOrder2 = inOrder(obj);

        inOrder2.verify(obj).wait(anyLong(), anyInt());
    }
}
```

**Comment by shivanisky** [comment]
Yes … but possibly this white space issue is unrelated to this PR. As I
think I’ve seen it in Expectations migration, maybe.

Will have to check it out. It’s happening in a specific case only.

**Comment by timtebeek** [comment]
Ah check! Thanks for looking into it; also let me know if you'd like me to take over. The way I see it there's a few options:

1. Counteract the left over whitespace in the recipe
2. Try to resolve the issue upstream somewhere
3. Adjust the tests to match the current output

All three could be a viable way to merge this PR, and don't necessarily rule out one another; I think folks would prefer to see these case migrated, even if there are excess newlines, for now. Then if the issue is fixed upstream we can revisit. Let me know what you find and which direction you want to go. :)

**Comment by shivanisky** [comment]
Hi Tim

I’m more than happy for you to take over. A bit tied up with work (java
migration hehe).

Any of those options look good. It seems the issue is related to the number
of lines before migration vs the number of lines in the new code. I
couldn’t see any issue in the recipe but I’m still new to this. Suspect
issue is in the framework.

Thanks

Shivani

**Review by timtebeek — APPROVED** [review]
Thanks a lot @shivanisky ! Great to see this verifications in order covered as well.

I explored the whitespace issue briefly, and noticed a missing semicolon, but that failed to resolve the issue. We'll accept that for now and can circle back later if needed. I don't think folks will stumble over this in the migration, so then it's better to see it covered.

**Pull request merged**